### PR TITLE
Use int-based hashmaps and enumerated algorithm values

### LINKED ISSUE #89: Hash algorithm enumerations are not used
## Short Description:

Nodes should use the enumerated hash algorithms.

## Details:

:warning: **Please note that the fix for this bug involves a breaking change** :warning: 

The proto definitions include an enumeration of the supported hash algorithms:

https://github.com/bom-squad/protobom/blob/4fc3791784be45015eb6a374db9d5972134f935f/api/sbom.proto#L153-L174

However, the node structure and the external reference is using a map of strings keyed with strings:

https://github.com/bom-squad/protobom/blob/4fc3791784be45015eb6a374db9d5972134f935f/api/sbom.proto#L28
https://github.com/bom-squad/protobom/blob/4fc3791784be45015eb6a374db9d5972134f935f/api/sbom.proto#L128

We should use the enumerated constants as keys in the hash entries. Having the value as strings opens up the use of protobom to all sorts of inconsistencies, case mixups and so on.

### PR DESCRIPTION
**Pull request by puerco**
> **:warning: DO NOT MERGE YET :warning:** 
> To be discussed in the Sep 6th '23 cohort meeting

This PR changes protobom to use integer-based maps in the hash lists to prevent inconsistencies in the ingested and serialized algorithm names.  

See issue #89 for more details on the problem this PR fixes.

This is the full list of updates this PR addresses:

- [x] Update protobuf
- [x] Regenerate go types from proto
- [x] Update node and nodelists objects and methods (and their respective tests)
- [x] Update serializers
- [x] Update unserializers
- [x] Update new alpha SPDX3 code
- [x] Regenerate the existing conformance golden samples

Closes #89

Signed-off-by: Adolfo García Veytia (Puerco) <puerco@chainguard.dev>

### CODE REVIEW

**Comment by jspeed-meyers** [comment]
> Aug 6th '23 

I think it's the Sep. 6th '23 meeting

**Comment by veramine** [comment]
Would it make sense to use a HashAlgorithm as the key instead of an int32?  One is just a typedef to the other but calling it a HashAlgorithm may help humans if protoc will let you do it easily

**Comment by puerco** [comment]
@veramine unfortunately you can't. protobuf3 does not allow enumerated values to be used as keys :/

**Comment by veramine** [comment]
alright, no worries

**Comment by veramine** [comment]
I'm in favor of this change

File: api/sbom.proto
Comment (veramine) [inline api/sbom.proto:1]: It doesn't really matter that much but do you need to re-number "identifiers" here?  Were you just targeting hashes and accidentally changed identifiers as well?

Comment (puerco) [reply api/sbom.proto:1]: Oops no sorry, I bumped the wrong field number. Good catch @veramine !
I've fixed it and repushed the branch.

**Review by veramine — APPROVED** [review]

**Comment by puerco** [comment]
Cohort emoji/vocally approved this change on the sep 6th meeting :+1:

**Pull request merged**